core/utils/xml.js

```diff
@@ -76,5 +84,29 @@ Blockly.utils.xml.textToDomDocument = function(text) {
  */
 Blockly.utils.xml.domToText = function(dom) {
   var oSerializer = new XMLSerializer();
-  return oSerializer.serializeToString(dom);
+  return Blockly.utils.xml.sanitizeText(oSerializer.serializeToString(dom));
+};
+
+/**
+ * Sanitizes a text string so that it can be serialized to XML. Specifically,
+ * this replaces control characters with their NCR (numerical character
+ * reference).
+ * @param {string} text The text to sanitize.
+ * @return {string} The sanitized text.
+ */
+Blockly.utils.xml.sanitizeText = function(text) {
+  if (!text.match(Blockly.utils.xml.INVALID_CONTROL_CHARS)) {
+    return text;
+  }
+  var match;
+  var newText = '';
+  while ((match = text.match(Blockly.utils.xml.INVALID_CONTROL_CHARS))) {
+    newText += text.substring(0, match.index);
+    newText += '&#x';
+    newText += text.charCodeAt(match.index).toString(16);
+    newText += ';';
+    text = text.substr(match.index + 1);
```

Comment:
"Is there a reason you used `substr` instead of `substring` here as `substr` technically doesn't seem to be part of the core ECMAScript language?"

Reply:
It looks like the ECMAScript people have backed off of that? I don't see any warning on the [MDN page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr) anymore. But I'm not sure.

Reply:
`String.prototype.substr` is in Appendix B, which is to say that it is required for web compatibility.  [Specifically](https://tc39.es/ecma262/multipage/additional-ecmascript-features-for-web-browsers.html#sec-additional-ecmascript-features-for-web-browsers):

> These features are not considered part of the core ECMAScript language. Programmers should not use or assume the existence of these features and behaviours when writing new ECMAScript code. ECMAScript implementations are discouraged from implementing these features unless the implementation is part of a web browser or is required to run the same legacy ECMAScript code that web browsers encounter.

I'm not sure why MDN would have removed their warning, though perhaps it is fair to say that on the one hand `substr` is for all practical purposes here to stay (even in non-browser engines like node.js) while on the other hand is not deprecated (because it does not officially exist at all).  🤷‍♂️